File: src/test/java/io/shiftleft/overflowdb/structure/specialized/gratefuldead/FollowedBy.java
package io.shiftleft.overflowdb.structure.specialized.gratefuldead;

import io.shiftleft.overflowdb.structure.*;

import java.util.*;

public class FollowedBy extends OverflowDbEdge {
  public static final String LABEL = "followedBy";
  public static final String WEIGHT = "weight";
  public static final HashSet<String> PROPERTY_KEYS = new HashSet<>(Arrays.asList(WEIGHT));

  public Integer weight() {
    return (Integer) property(WEIGHT).value();
  }

  public FollowedBy(TinkerGraph graph, VertexRef<OverflowDbNode> outVertex, VertexRef<OverflowDbNode> inVertex) {
    super(graph, LABEL, outVertex, inVertex, PROPERTY_KEYS);
  }

  public static final EdgeLayoutInformation layoutInformation = new EdgeLayoutInformation(LABEL, PROPERTY_KEYS);

  public static OverflowElementFactory.ForEdge<FollowedBy> factory = new OverflowElementFactory.ForEdge<FollowedBy>() {
    @Override
    public String forLabel() {
      return FollowedBy.LABEL;
    }

    @Override
    public FollowedBy createEdge(TinkerGraph graph, VertexRef outVertex, VertexRef inVertex) {
      return new FollowedBy(graph, outVertex, inVertex);
    }
  };
}

File: traversal/src/test/scala/io/shiftleft/overflowdb/traversal/testdomains/gratefuldead/FollowedBy.scala
package io.shiftleft.overflowdb.traversal.testdomains.gratefuldead

import io.shiftleft.overflowdb._
import scala.jdk.CollectionConverters._

class FollowedBy(graph: OdbGraph, outVertex: NodeRef[SongDb], inVertex: NodeRef[SongDb])
  extends OdbEdge(graph, FollowedBy.Label, outVertex, inVertex, FollowedBy.PropertyNames.all.asJava)

object FollowedBy {
  val Label = "followedBy"

  object Properties {
    val Weight = "weight"
  }

  object PropertyNames {
    val Weight = "weight"
    val all: Set[String] = Set(Weight)
  }

  val layoutInformation = new EdgeLayoutInformation(Label, PropertyNames.all.asJava)
  var factory: EdgeFactory[FollowedBy] = new EdgeFactory[FollowedBy] {
    override def forLabel(): String = FollowedBy.Label

    override def createEdge(graph: OdbGraph, outNode: NodeRef[OdbNode], inNode: NodeRef[OdbNode]): FollowedBy =
      new FollowedBy(graph, outNode.asInstanceOf[NodeRef[SongDb]], inNode.asInstanceOf[NodeRef[SongDb]])
  }
}